File: src/tetris/core/game_state.py
from .active_piece import ActivePiece
from .bag import SevenBag
from .board import Board
from .lock_event import LockEvent
from .tetromino import PieceType, Tetromino


_T_PIVOTS = {
    0: (1, 1),
    1: (0, 1),
    2: (1, 0),
    3: (1, 1),
}


class GameState:
    def __init__(
        self,
        seed=None,
        width=10,
        visible_height=20,
        hidden_rows=2,
        next_count=5,
    ):
        if not isinstance(next_count, int) or isinstance(next_count, bool) or next_count < 0:
            raise ValueError("next_count must be a nonnegative integer")

        self.board = Board(width, visible_height, hidden_rows)
        self.bag = SevenBag(seed)
        self.next_count = next_count
        self._next_queue = [next(self.bag) for _ in range(next_count)]
        self.active = None
        self.hold = None
        self.hold_used = False
        self.game_over = False
        self.lines = 0
        self.combo = 0
        self.back_to_back_active = False
        self.pieces_locked = 0
        self.last_lock_event: LockEvent | None = None
        self._last_rotation_successful = False
        self.spawn()

    @property
    def next_pieces(self) -> tuple[PieceType, ...]:
        return tuple(self._next_queue)

    def spawn(self):
        if self._next_queue:
            kind = self._next_queue.pop(0)
            self._next_queue.append(next(self.bag))
        else:
            kind = next(self.bag)

        self.active = ActivePiece(kind, (self.board.width - 4) // 2, 0)
        self._last_rotation_successful = False
        if not self.board.can_place(
            Tetromino(self.active.kind, self.active.cells()),
            (self.active.x, self.active.y),
        ):
            self.game_over = True

    def _can(self, piece):
        return self.board.can_place(
            Tetromino(piece.kind, piece.cells()),
            (piece.x, piece.y),
        )

    def move(self, dx, dy=0):
        candidate = ActivePiece(
            self.active.kind,
            self.active.x + dx,
            self.active.y + dy,
            self.active.rotation,
        )
        if self._can(candidate):
            self.active = candidate
            if dx != 0:
                self._last_rotation_successful = False
            return True
        return False

    def rotate(self, direction=1):
        candidate = ActivePiece(
            self.active.kind,
            self.active.x,
            self.active.y,
            self.active.rotation + direction,
        )
        if self._can(candidate):
            self.active = candidate
            self._last_rotation_successful = True
            return True
        for dx in (-1, 1, -2, 2):
            candidate.x = self.active.x + dx
            if self._can(candidate):
                self.active = candidate
                self._last_rotation_successful = True
                return True
        return False

    def hard_drop(self):
        distance = 0
        while self.move(0, 1):
            distance += 1
        self.lock()
        return distance

    def lock(self):
        locked_piece = self.active
        t_spin = self._is_t_spin(locked_piece)
        self.board.lock(
            Tetromino(locked_piece.kind, locked_piece.cells()),
            (locked_piece.x, locked_piece.y),
        )
        cleared_lines = self.board.clear_full_rows()
        self.lines += cleared_lines
        self.pieces_locked += 1

        if cleared_lines > 0:
            self.combo += 1
        else:
            self.combo = 0

        difficult_clear = cleared_lines == 4 or (t_spin and cleared_lines > 0)
        back_to_back = difficult_clear and self.back_to_back_active
        if difficult_clear:
            self.back_to_back_active = True
        elif cleared_lines > 0:
            self.back_to_back_active = False

        perfect_clear = cleared_lines > 0 and not any(self.board.cells())
        self.last_lock_event = LockEvent(
            lock_id=self.pieces_locked,
            piece=locked_piece.kind,
            lines=cleared_lines,
            t_spin=t_spin,
            combo=self.combo,
            back_to_back=back_to_back,
            perfect_clear=perfect_clear,
        )

        self.hold_used = False
        self.spawn()
        return self.last_lock_event

    def hold_piece(self):
        if self.hold_used:
            return False
        old = self.hold
        self.hold = self.active.kind
        self.hold_used = True
        self._last_rotation_successful = False
        if old is None:
            self.spawn()
        else:
            self.active = ActivePiece(old, (self.board.width - 4) // 2, 0)
        if not self._can(self.active):
            self.game_over = True
        return True

    def _is_t_spin(self, piece: ActivePiece) -> bool:
        if piece.kind != PieceType.T or not self._last_rotation_successful:
            return False

        pivot_x, pivot_y = _T_PIVOTS[piece.rotation % 4]
        center_x = piece.x + pivot_x
        center_y = piece.y + pivot_y
        occupied_corners = 0
        for dx, dy in ((-1, -1), (1, -1), (-1, 1), (1, 1)):
            x = center_x + dx
            y = center_y + dy
            if x < 0 or x >= self.board.width or y < 0 or y >= self.board.height:
                occupied_corners += 1
            elif self.board.occupied(x, y):
                occupied_corners += 1
        return occupied_corners >= 3

File: core/src/game_state.cpp
#include "kadoka/tetris/game_state.hpp"

#include "kadoka/tetris/combat_rules.hpp"

#include <utility>

namespace kadoka::tetris {

GameState::GameState(
    std::uint64_t seed,
    int width,
    int visible_height,
    int hidden_rows,
    std::size_t next_count)
    : board_(width, visible_height, hidden_rows),
      bag_(seed),
      next_count_(next_count) {
    next_queue_.reserve(next_count_);
    for (std::size_t i = 0; i < next_count_; ++i) {
        next_queue_.push_back(bag_.next());
    }
    spawn();
}

bool GameState::move(int dx, int dy) {
    ActivePiece candidate = active_;
    candidate.x += dx;
    candidate.y += dy;
    if (!can_place(candidate)) {
        return false;
    }

    active_ = candidate;
    if (dx != 0) {
        last_rotation_successful_ = false;
    }
    return true;
}

bool GameState::rotate(int direction) {
    ActivePiece candidate = active_;
    candidate.rotation += direction;
    if (can_place(candidate)) {
        active_ = candidate;
        last_rotation_successful_ = true;
        return true;
    }

    for (const int dx : {-1, 1, -2, 2}) {
        candidate.x = active_.x + dx;
        if (can_place(candidate)) {
            active_ = candidate;
            last_rotation_successful_ = true;
            return true;
        }
    }
    return false;
}

int GameState::hard_drop() {
    int distance = 0;
    while (move(0, 1)) {
        ++distance;
    }
    (void)lock();
    return distance;
}

LockEvent GameState::lock() {
    const ActivePiece locked_piece = active_;
    const bool t_spin = is_t_spin(locked_piece);

    board_.lock(as_tetromino(locked_piece), {locked_piece.x, locked_piece.y});
    const int cleared_lines = board_.clear_full_rows();
    lines_ += cleared_lines;
    ++pieces_locked_;

    if (cleared_lines > 0) {
        ++combo_;
    } else {
        combo_ = 0;
    }

    const bool difficult_clear = cleared_lines == 4 || (t_spin && cleared_lines > 0);
    const bool back_to_back = difficult_clear && back_to_back_active_;
    if (difficult_clear) {
        back_to_back_active_ = true;
    } else if (cleared_lines > 0) {
        back_to_back_active_ = false;
    }

    const bool perfect_clear = cleared_lines > 0 && board_.occupied_cells().empty();
    LockEvent event{
        pieces_locked_,
        locked_piece.kind,
        cleared_lines,
        t_spin,
        combo_,
        back_to_back,
        perfect_clear,
    };
    last_lock_event_ = event;

    hold_used_ = false;
    spawn();
    return event;
}

bool GameState::add_garbage(const std::vector<int>& hole_columns) {
    const bool overflow = board_.add_garbage(hole_columns);
    last_rotation_successful_ = false;
    if (overflow || !can_place(active_)) {
        game_over_ = true;
    }
    return overflow;
}

bool GameState::hold_piece() {
    if (hold_used_) {
        return false;
    }

    const std::optional<PieceType> previous_hold = hold_;
    hold_ = active_.kind;
    hold_used_ = true;
    last_rotation_successful_ = false;

    if (!previous_hold.has_value()) {
        spawn();
    } else {
        active_ = ActivePiece{
            *previous_hold,
            (board_.width() - 4) / 2,
            0,
            0,
        };
    }

    if (!can_place(active_)) {
        game_over_ = true;
    }
    return true;
}

void GameState::spawn() {
    PieceType kind{};
    if (!next_queue_.empty()) {
        kind = next_queue_.front();
        next_queue_.erase(next_queue_.begin());
        next_queue_.push_back(bag_.next());
    } else {
        kind = bag_.next();
    }

    active_ = ActivePiece{
        kind,
        (board_.width() - 4) / 2,
        0,
        0,
    };
    last_rotation_successful_ = false;
    if (!can_place(active_)) {
        game_over_ = true;
    }
}

bool GameState::can_place(const ActivePiece& piece) const noexcept {
    return board_.can_place(as_tetromino(piece), {piece.x, piece.y});
}

bool GameState::is_t_spin(const ActivePiece& piece) const {
    return is_t_spin_placement(
        piece.kind,
        piece.rotation,
        piece.x,
        piece.y,
        board_.width(),
        board_.height(),
        [this](int x, int y) { return board_.occupied(x, y); },
        last_rotation_successful_);
}

Tetromino GameState::as_tetromino(const ActivePiece& piece) {
    return Tetromino{piece.kind, piece.cells()};
}

}  // namespace kadoka::tetris
